### PR TITLE
Check password strength

### LINKED ISSUE #4: Password Strength
Write a certain function that checks certain data about the received password and returns whether it is correct

### PR DESCRIPTION
**Pull request by FridaPshitik**
Write a function that checks that each received password will be correct according to certain criteria

Closes #4

### CODE REVIEW

**Comment by bdsoha** [comment]
- [ ] Move your code into a sub-directory so that it shouldn't conflict with others.
- [ ] Add a readme to the sub-directory explaining how to install/use your code.

**Comment by bdsoha** [comment]
Please change the directory name to something more significant, like `password-strength`.

**Comment by bdsoha** [comment]
I will begin the code review now.

File: password-strength/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+Password Strength Checker
```
Comment (bdsoha) [inline password-strength/README.md:1]:
```suggestion
# Password Strength Checker
```

File: password-strength/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+- Submit bug reports or feature requests via the issue tracker on GitHub.
+- Fork the repository and create a branch for your changes.
+- Follow the project's coding guidelines and standards.
+- Submit a pull request with your changes and a description of what you've done.
+
+If you have any questions or comments, please feel free to contact the project owner at FridaP@skyvar.co.il or through the GitHub repository.
+
+## License
+
+This project is licensed under the MIT License - see the LICENSE file for details.
```
Comment (bdsoha) [inline password-strength/README.md:46]: How and why are you licensing your code?
```suggestion
```

File: password-strength/dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+FROM node:16.19.1-buster
```
Comment (bdsoha) [inline password-strength/dockerfile:2]:
```suggestion
FROM node:16.19.1-buster
```

File: password-strength/dockerfile [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+WORKDIR /passwordString
```
Comment (bdsoha) [inline password-strength/dockerfile]: Do you need to set a `WORKDIR`?

**Review by bdsoha — CHANGES_REQUESTED** [review]

File: password-strength/app.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+
+function PasswordStrength(password) {
+
+    if (isLowerCharacter(password) && isUpperCharacter(password)
+        && isDigit(password) && isSpecialCharacter(password)
+        && isLength(password) && isRepeatMoreThanTwice(password)
+        && isSequencesOfThree(password))
+        return true;
+    return false;
```
Comment (bdsoha) [inline password-strength/app.js:9]:
```suggestion
    return isLowerCharacter(password) 
        && isUpperCharacter(password)
        && isDigit(password)
        && isSpecialCharacter(password)
        && isLength(password)
        && isRepeatMoreThanTwice(password)
        && isSequencesOfThree(password)
```

**Review by bdsoha — CHANGES_REQUESTED** [review]

**Comment by bdsoha** [comment]
### Few Questions
- What are the any advantages and disadvantages of using `regex`?
- How would OOP assist in abstracting the code?
- Can you test the current implementation at the _unit_ level?

**Comment by FridaPshitik** [comment]
Is this the OOP usage you meant?

**Comment by bdsoha** [comment]
> * How would OOP assist in abstracting the code?

I didn't ask you to implement, I just asked you *How would you*.
It is important to pay attention to the small details, it saves time, and ⌚is 💲

**Comment by FridaPshitik** [comment]
Using OOP you can put all password handling functions under one roof
In my opinion the current task does not have a critical need for OOP

**Comment by bdsoha** [comment]
> Using OOP you can put all password handling functions under one roof In my opinion the current task does not have a critical need for OOP

You opinion is noted, IMO, there is a very strong case for OOP.
Using OOP you would be able to:
- Remove all `if` statements.
- Abstractly handle character conversions.
- Unit test specific rules.
- Dynamically disable rule checking.
- Log flow messages *(debug)*.


```js
class AbstractBaseRule {
    _handle(password) {
        if (this.check(password)) {
            // ...
        }
    }

    check(password) {
        throw new Error('Abstract non-implemented method');
    }
}
```

**Comment by FridaPshitik** [comment]
Do you mean all the IF's were in one function

**Comment by bdsoha** [comment]
> Do you mean all the IF's were in one function

Based on the code I provided the inherited rule would look like this:

```js
class HasLowerCase extends AbstractBaseRule {
    check(password) {
        return /[a-z]/.test(password)
    }
}
```

This can be individually tested and extendable.

**Comment by FridaPshitik** [comment]
REJEX is a useful mechanism for identifying and filtering different types of data
The advantages of REJEX are:

1. Time reduction - the use of REJEX may save a lot of time in the implementation and filtering of the data, so that the user can act faster and save valuable time.
2. Accuracy - REJEX makes it possible to identify the data in a more efficient way and thus avoid mistakes and also get more accurate results.
3. Flexibility - REJEX can be adapted to a wide variety of data,

The disadvantages of REJEX are:

1. REJEX must be correct in the implementation and definition of the templates, so if the template is wrong or inaccurate, the user may miss important data or get inaccurate results.
2. REJEX commands are more complex to learn

Since the pros outweigh the cons, I'm in the middle of switching to using rejex

**Comment by FridaPshitik** [comment]
class HasRepeatMoreThanTwice extends AbstractBaseRule {
    check(password) {
        let pass = /([a-zA-Z0-9])\1{2,}/
        return pass.test(password)
    }
}

Do you know why this function is not working?
I try many ways and nothing works

**Comment by bdsoha** [comment]
> class HasRepeatMoreThanTwice extends AbstractBaseRule { check(password) { let pass = /([a-zA-Z0-9])\1{2,}/ return pass.test(password) } }
> 
> Do you know why this function is not working? I try many ways and nothing works

Assuming the `AbstractBaseRule` exists:
```js
new HasRepeatMoreThanTwice().check('someteststring')
```

**Comment by bdsoha** [comment]
Don't change the code again.
Pay attention to what was written [here](https://github.com/skyvar/registration/pull/26#issuecomment-1475134458).

This assignment is a learning experience, the fact that I ask you a question doesn't mean "go implement this... or change that...".

**Comment by FridaPshitik** [comment]
The disadvantages and advantages of using regex
I have already registered
You explained to me how OOP will help simplify the code
And in the usual way that I wrote down, it can be more complicated to check edge cases
But now the code works in all modes I fixed a small mistake

**Comment by bdsoha** [comment]
**🎉 Congrats on your first merged PR.**

You will soon be added as a collaborator to this repo.
In the meantime, please prepare a short presentation *(in English)* describing:

- The task in general.
- Things you learned along the way.
- Tips for other team-members for future tasks.

**Comment by bdsoha** [comment]
You have been sent an email invitation to the [SkyVar Organization](https://github.com/skyvar).
Please open your cooperate email and finish the registration process.

**Pull request merged**